### ISSUE TITLE
❓ [QUESTION]

### ISSUE THREAD

**Issue by truzmeto** [e3nn/e3nn#192]
Hi all, 
thanks for making this repo public, great work!
I was trying to do equivariance test for autoencoder type network for for a small peptide(tri-peptide) with voxel convolution. Somehow can't get it working for vector fields and was wondering if I'm doing it right. 
My input and the output are represented as Gaussian densities in a 3D grid. Latent representation is compressed via series of layers into 36 channels, where each channel is 1x1x1 voxel. It has 22 scalars (L=0, 22 channels), 3 vectors (L=1, 9 channels), and 1 tensor (L=2, 5 channels). Is it correct to assume that initial latent[0:22] are L0, then following latent[22:31] channels belong to L1, and remaining latent[31:36] are L2?  I'm using a version of code from last year,  perhaps very similar to current master branch.  So, when I apply trained network on rotated versions of input, the output rotate very similarly. But latent vectors don't, also their length changes as I do rotation. Below I attached the simple  model I'm using, sorry it's not minimal code, but very easy to read. Also, attached a video of my rotation equivariance to better  illustrate the problem I'm having. In a video, 3 vectors in the middle extracted from latent -->  vec = latent[22:31]. Then reshaped to xyz to plot. 
Question is: Why output is equivariant wrt rotations but not latent vectors? 
Thansk! 

`
import sys
import os

import torch
import torch.nn as nn
import torch.nn.functional as F

from e3nn.non_linearities.rescaled_act import sigmoid, swish, tanh
from e3nn.non_linearities.norm_activation import NormActivation
from e3nn.batchnorm import BatchNorm

sys.path.append(os.path.join(os.path.dirname(__file__), "../.."))
from src.SE3Model.convolution import Convolution

def ConvBlock(Rs_in, Rs_out, size, stride, fpix):
    return nn.Sequential(
        Convolution(Rs_in, Rs_out, size = size, stride = stride, padding = size//2, bias=None, fuzzy_pixels = fpix),
        NormActivation(Rs_out, swish, normalization = 'norm'),
    )

def ConvTBlock(Rs_in, Rs_out, size, fpix):
    return nn.Sequential(
        Convolution(Rs_in, Rs_out, size=size, stride=2, padding=size//2, bias=None, output_padding=1, transpose=True, fuzzy_pixels = fpix),
        #NormActivation(Rs_out, swish, normalization = 'componenet'),
    )

class VNet(nn.Module):
    def __init__(self, size, n_reps, inp_channels=11):
        super(VNet, self).__init__()
        
        Rs_in = [(inp_channels,0)]
        Rs_out =  [(inp_channels,0)]
        m = inp_channels

        n_vec, n_ten = n_reps
        RsSca = [0]; RsVec = [1] * n_vec; RsTen = [2] * n_ten
        Rs = RsSca * m
        Rs1 = RsVec + RsTen
        fp = True  #option to add noise to conv kernels
        
        #down
        self.dw1 = ConvBlock(Rs_in,          Rs +  Rs1, size=size, stride = 2, fpix=fp)
        self.dw2 = ConvBlock(   Rs +  Rs1, 2*Rs +  Rs1, size=size, stride = 2, fpix=fp)
        self.dw3 = ConvBlock(2* Rs +  Rs1, 2*Rs +  Rs1, size=size, stride = 2, fpix=fp)
        self.dw4 = ConvBlock(2* Rs +  Rs1, 2*Rs +  Rs1, size=size, stride = 2, fpix=fp)
        
        #up
        self.up1 = ConvTBlock( 2*Rs + Rs1, 2*Rs + Rs1, size=size, fpix=fp)
        self.cd1 = ConvBlock(  2*Rs + Rs1, 2*Rs + Rs1,  size=size, stride = 1, fpix=fp)
        self.up2 = ConvTBlock( 2*Rs + Rs1, 2*Rs + Rs1,  size=size, fpix=fp)
        self.cd2 = ConvBlock(  2*Rs + Rs1, 2*Rs + Rs1,  size=size, stride = 1, fpix=fp)
        self.up3 = ConvTBlock( 2*Rs + Rs1, 2*Rs + Rs1,  size=size, fpix=fp)
        self.cd3 = ConvBlock(  2*Rs + Rs1,   Rs + Rs1,  size=size, stride = 1, fpix=fp)
        self.up4 = ConvTBlock(   Rs + Rs1,   Rs + Rs1,  size=size, fpix=fp)
        self.cd4 = ConvBlock(    Rs + Rs1,   Rs_out,    size=size, stride = 1, fpix=fp)

        
    def forward(self, x):
        # Down sampling
        dw1 = self.dw1(x) 
        dw2 = self.dw2(dw1) 
        dw3 = self.dw3(dw2) 
        dw4 = self.dw4(dw3) 
        
        up1 = self.up1(dw4) 
        cd1 = self.cd1(up1)
        up2 = self.up2(cd1) 
        cd2 = self.cd2(up2) 
        up3 = self.up3(cd2) 
        cd3 = self.cd3(up3) 
        up4 = self.up4(cd3) 
        cd4 = self.cd4(up4)
        
        return cd4, dw4

    
if __name__ == "__main__":

  inp_size = 16
  inp_channels = 11
  n_reps = [3,1] # n_vectors (L=1), n_tensors (L=2)
  
  k_size = 3
  
  x = torch.Tensor(2, inp_size, inp_size, inp_size, inp_channels).cuda()
  #x.to(device)
  print("Input size: {}".format(x.size()))
  
  model = VNet(size = k_size, n_reps = n_reps).cuda()
  
  out, latent = model(x)
  print("latent size: {}".format(latent.size()))
  print("output size: {}".format(out.size()))
 
`

https://user-images.githubusercontent.com/18493775/106665964-f79ff880-6574-11eb-8c89-bbc74b6c6c97.mp4

**Comment by blondegeek** [comment]
Hi Talant,

"vectors" (L=1) in e3nn are in the order [y, z, x] -- the same convention
as real spherical harmonics. Did you ensure to permute the output indices
to match the Cartesian convention [x, y, z] when you compared the model
output to your ground truth?

Best,
Tess

**Comment by truzmeto** [comment]
Hi Tess, 

yes I permuted the output like this: 
`
z = torch.einsum('txyzi->tixyz', z).squeeze() #unpermute
`
Ic, I'll try to transform [y,z,x] --> [x,y,z] and check if it works.
Thanks!
Talant

**Comment by blondegeek** [comment]
Hi Talant

Specifically you should only transform the representation index (the i
index) which carries the vector information -- not the spatial indices -
those are already given in cartesian coordinates. you will need a
permutation matrix like [[0, 0, 1], [1, 0, 0], [0, 1, 0]].

Best,
Tess

**Comment by truzmeto** [comment]
Hi Tess, yes I understood that L1 components need to be reordered. I tried that, it didn't help. 
I decided to post even more simple code here in order to better understand how this vector representation works.
This time I use a simple 1 layer convolution taken from unit test here: 
[convUnitTestLink]( https://github.com/e3nn/e3nn/blob/447ccb253061a50b29f3a05c6eeffba34cca2c14/tests/image/convolution_test.py#L10)

Input is Rs_in = [0] or 16x16x16 scalar. Output is Rs_out = [0,1] 1 scalar 14x14x14, and 1 vector field 14x14x14, 3 channels.
Input has 1 non-zero voxel at center and other one away from center, just two non-zero voxels. Reordering of L1 components is done in line 96, where I simple do [L1y, L1z, L1x] --> [L1x, L1y, L1z]  . The way vector field rotates is a bit strange. It does seem to have global equivariace, but lack local. Is it expected behaviour, or am I doing smth wrong here? 

`
import torch
import torch.nn as nn
import numpy as np
import pyvista as pv
from scipy import ndimage
from e3nn import rs
from e3nn.image.convolution import Convolution

class Simple(nn.Module):
    def __init__(self, fuzzy_pixels):
        super(Simple, self).__init__()
        size = 3
        self.f = torch.nn.Sequential( Convolution(Rs_in=[0], Rs_out=[0, 1], size=size,  steps=(1., 1., 1.), fuzzy_pixels=fuzzy_pixels),
        )
    def forward(self, x):
        out = self.f(x)
        return out

def rotate(inp, rotation_angle):    
    inp = ndimage.interpolation.rotate(inp,
                                       angle = rotation_angle,
                                       axes=(2,3),
                                       reshape=False,
                                       order=1,
                                       mode= 'nearest',#'constant',
                                       cval=0.0)
    return inp


def VoxPositions(dim, res):
    pos = np.empty(shape=(dim*dim*dim,3))
    l = 0
    for i in range(dim):
        for j in range(dim):
            for k in range(dim):
                pos[l, 0] = i*res + res/2
                pos[l, 1] = j*res + res/2
                pos[l, 2] = k*res + res/2
                l = l + 1 
    return pos

if __name__=='__main__':

    dim = 16;  chans = 1
    torch.manual_seed(1000)

    # get input
    inp = torch.zeros(1, chans, dim, dim, dim)
    inp[:, :, dim//2, dim//2, dim//2] = 1.0
    inp[:, :, dim//2, 1, dim//2] = 1.0
    model = Simple(fuzzy_pixels=True)
    
    # plotting setup
    pl = pv.Plotter(shape=(1, 3))
    pl.open_movie('Simple.mp4')
    fs = 12 #text font size
    cents = VoxPositions(dim=dim-2, res = 1.0)

    for i in range(180):
        inpR = rotate(inp.numpy(), rotation_angle= i*2.0)
        inpR = torch.from_numpy(inpR).float()#.to('cuda')
        inpR = torch.einsum('tixyz->txyzi', inpR) #permute
                
        model.eval()
        outR = model(inpR)
        outR = torch.einsum('txyzi->tixyz', outR) #unpermute
        inpR = torch.einsum('txyzi->tixyz', inpR) #unpermute

        OutSca = outR[0, 0, :, :, :].detach().numpy()
        OutVec = outR[0, 1:4, :, :, :]
        OutVec = OutVec.flatten(1).detach().numpy()

        vec = np.array([OutVec[2], OutVec[0], OutVec[1]]).T
             
        text = "angle = " + str(2*i)
        pl.subplot(0, 0);  
        pl.add_text("Input", position = 'lower_left', font_size = fs)
        pl.add_text(text, position = 'upper_left', font_size = fs)
        pl.add_volume(inpR[0][0].detach().numpy(), cmap = "viridis_r",
                      opacity = "linear", show_scalar_bar=False)

        pl.subplot(0, 1);  
        pl.add_text("Out Vector", position = 'lower_left', font_size = fs)
        pl.add_arrows(cents, vec, mag=20, show_scalar_bar=False)
        
        pl.subplot(0, 2);  
        pl.add_text("Output", position = 'lower_left', font_size = fs)
        OutSca[OutSca < 0.01] = 0.0
        pl.add_volume(OutSca, cmap = "viridis_r", show_scalar_bar=False)
        #pl.add_axes()

        if i == 0 :
            pl.show(auto_close=False)
          
        pl.write_frame()
        pl.clear()

    pl.close()`

Video is attached.
Thank You so much!

https://user-images.githubusercontent.com/18493775/107169484-f19e8300-698b-11eb-8cec-d5e38dd8fc2c.mp4

**Comment by mariogeiger** [comment]
Hi, Looking at the video it's hard to say if it's correct or not.
What you see can be due to
- pixel interpolation
- border effects (your second pixel is one pixel away from the border)

Note that your two pixels dont interact since they are at distance 7 and your kernel is of size 3

**Comment by truzmeto** [comment]
Hi Mario, 

thanks for your feedback!

- Yes, rotation in grid uses interpolation, and alters the input..
- Border effects are hard to avoid in grid, I agree

So, I made a more robust input with shorter range interactions, and larger dimension. It worked, vector field rotations make much better sense.

Thanks again!
Talant 

https://user-images.githubusercontent.com/18493775/107254198-1420b280-6a05-11eb-9243-d22981c9ec69.mp4

**Comment by mariogeiger** [comment]
Nice!

**Comment by mariogeiger** [comment]
Note that in case you switch to the last version of the code, the L=1 representation now match with the cartesian coordinates

**Comment by blondegeek** [comment]
Hi @truzmeto,
Did our responses answer your questions? If so, I'd like to close this issue.
Best!
Tess

**Comment by truzmeto** [comment]
Hi @blondegeek , yes your responses were super helpful! Thank you!

**Comment by blondegeek** [comment]
You are most welcome! Thank you for your questions.
